app/id_generation.py

```diff
@@ -2,41 +2,142 @@
-            + self.generate_three_digit_code()
+    def __init__(self, parameters):
+        self.parameters = parameters
+
+        record_already_exist, ID = self.dedupe_for_users()
+        if record_already_exist:
+            self.id = ID
+
+        else:
+            counter = 1000
+            while counter > 0:
+                id = (
+                    self.get_class_code()
+                    + self.get_jnv_code()
+                    + self.generate_three_digit_code()
+                )
+
+                if not self.check_for_duplicate_ID(id):
+                    self.id = id
+                    break
+                counter -= 1
+
+    def dedupe_for_users(self):
+
+        # check if the keys used for deduping exist in the request data
+        for key in ["grade", "date_of_birth", "school_name", "gender", "category"]:
+            if (
+                key not in self.parameters
+                or self.parameters[key] == ""
+                or self.parameters[key] is None
+            ):
+                raise HTTPException(
+                    status_code=400,
+                    detail="{} is not part of the request data".format(key),
+                )
+
+        does_user_already_exist = user_router.get_users(
+            build_request(
+                query_params={
+                    "date_of_birth": self.parameters["date_of_birth"],
+                    "gender": self.parameters["gender"],
+                }
+            )
         )
+        if len(does_user_already_exist) == 0:
+            # if user does not exist, go ahead with ID generation
+            return [False, ""]
+
+        else:
+            # if user does exist, check corresponding student attributes
+            for user in does_user_already_exist:
+                does_student_already_exist = student_router.get_students(
+                    build_request(
+                        query_params={
+                            "user_id": user["id"],
+                            "category": self.parameters["category"],
+                            "grade": self.parameters["grade"],
+                        }
+                    )
+                )
+
+                if len(does_student_already_exist) == 0:
+                    # if user is found, but a matching student is not found, then go ahead with ID generation
+                    return [False, ""]
+
+                else:
+                    # first, get the school ID based on the school name given in the request
+                    school_id_response = school.get_school(
+                        build_request(
+                            query_params={"name": self.parameters["school_name"]}
+                        )
+                    )
+                    if is_response_valid(
+                        school_id_response, "School ID could not be retrieved"
+                    ):
+                        school_id = is_response_empty(school_id_response.json())[0]
```

Comment:
`is_response_empty` seems like a weird function. if the response is indeed empty, it will return `[]` , and then we can't do `[0]` for it.